datafusion/common/src/scalar.rs

```diff
@@ -2156,30 +2157,58 @@ impl ScalarValue {
     ///
     /// assert_eq!(scalar_vec, expected);
     /// ```
-    pub fn convert_array_to_scalar_vec(array: &dyn Array) -> Result<Vec<Vec<Self>>> {
-        let mut scalars = Vec::with_capacity(array.len());
+    pub fn convert_list_array_to_scalar_vec(array: &dyn Array) -> Result<Vec<Vec<Self>>> {
+        let mut scalars_vec = Vec::with_capacity(array.len());
+
+        if as_list_array(array).is_ok() {
```

Comment:
In general, I think it would be better to avoid ignoring errors (which is what `is_ok()` does as each error requires at least one string allocation. This particular usage isn't likely performance critical but the pattern is not great and we have had trouble with it in the past.



It makes sense to me that we would need two different functions for Arrays to ScalarValue (as the return type of the ListArray needs to be different). 

I think it would be less surprising (and thus less error prone) if this function simply returned an error if the input was not a list array

Reply:
I think it would be nice to change both functions to ensure only ListArray is accepted in `convert_list_array_to_scalar_vec ` and non-ListArray is accepted in `convert_non_list_array_to_scalar_vec `.